### PR TITLE
Add test for aggregated revoked HTLC claim on anchors channel

### PR DESCRIPTION
**Pull request by wpaulino**
This PR adds a new test to cover the aggregated revoked HTLC claim edge case for channels featuring anchors outputs.

Along the way, we also fix a few issues to allow the anchors build to run, and we start running it in CI.

Fixes https://github.com/lightningdevkit/rust-lightning/issues/1905.

### CODE REVIEW

File: lightning/src/chain/onchaintx.rs [hunk 1]
```diff
@@ -72,18 +72,23 @@ impl OnchainEventEntry {
...
 	Claim {
 		package_id: PackageID,
 	},
-	/// Claim tx aggregate multiple claimable outpoints. One of the outpoint may be claimed by a counterparty party tx.
-	/// In this case, we need to drop the outpoint and regenerate a new claim tx. By safety, we keep tracking
-	/// the outpoint to be sure to resurect it back to the claim tx if reorgs happen.
+	/// The counterparty has claimed an outpoint from one of our pending requests through a
+	/// different transaction than ours. If our transaction was attempting to claim multiple
+	/// outputs, we need to drop the outpoint claimed by the counterparty and regenerate a new claim
+	/// transaction for ourselves. We keep tracking, separately, the outpoint claimed by the
```
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:89]: Can precise the duration of the counterparty claimed outpoint tracking, `ANTI_REORG_DELAY` iirc.

File: lightning/src/chain/onchaintx.rs [hunk 2]
```diff
@@ -248,15 +252,17 @@ pub struct OnchainTxHandler<ChannelSigner: WriteableEcdsaChannelSigner> {
...
-	// Value is (pending claim request identifier, confirmation_block), identifier
-	// is txid of the initial claiming transaction and is immutable until outpoint is
-	// post-anti-reorg-delay solved, confirmaiton_block is used to erase entry if
-	// block with output gets disconnected.
+	pending_claim_events: Vec<(PackageID, ClaimEvent)>,
+
+	// Used to link outpoints claimed in a connected block to a pending claim request. The keys
+	// represent the outpoints that our `ChannelMonitor` has detected we have keys/scripts to claim.
+	// The values track the pending claim request identifier and the initial confirmation block
+	// height, and are immutable until the outpoint is has enough confirmations to meet our
```
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:263]: s/ is has/has/g ?

**Review by ariard — COMMENTED** [review]
thanks for the additional test coverage, gonna look more

File: lightning/src/chain/onchaintx.rs [hunk 3]
```diff
@@ -248,15 +252,17 @@ pub struct OnchainTxHandler<ChannelSigner: WriteableEcdsaChannelSigner> {
...
 	#[cfg(anchors)]
-	pending_claim_events: HashMap<PackageID, ClaimEvent>,
-
-	// Used to link outpoints claimed in a connected block to a pending claim request.
-	// Key is outpoint than monitor parsing has detected we have keys/scripts to claim
-	// Value is (pending claim request identifier, confirmation_block), identifier
-	// is txid of the initial claiming transaction and is immutable until outpoint is
-	// post-anti-reorg-delay solved, confirmaiton_block is used to erase entry if
-	// block with output gets disconnected.
+	pending_claim_events: Vec<(PackageID, ClaimEvent)>,
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/onchaintx.rs:258]: I still think it's 483 as we shouldn't aggregate in a single `pending_claim_events` offered outputs (timeout path for us) and received outputs (preimage path for us). The runtime cost reasoning sounds correct, it's like at worst 482 `pending_claim_requests` entries + 1 `pending_claim_events` iiuc.

**Review by ariard — COMMENTED** [review]
Thanks for the great test!

File: lightning/src/chain/onchaintx.rs [hunk 4]
```diff
@@ -248,15 +252,17 @@ pub struct OnchainTxHandler<ChannelSigner: WriteableEcdsaChannelSigner> {
 	pub(crate) pending_claim_requests: HashMap<PackageID, PackageTemplate>,
 	#[cfg(not(test))]
 	pending_claim_requests: HashMap<PackageID, PackageTemplate>,
+
+	// Used to track external events that need to be forwarded to the `ChainMonitor`.
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/onchaintx.rs:252]: Yeah not public, though still think those interfaces might more exposed a day for dual-funding/splicing support, if `OnchainTxHandler` becomes its own trait :)

File: lightning/src/chain/onchaintx.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/onchaintx.rs:254]: Can document the above reasoning near `pending_claim_events`.

File: lightning/src/chain/onchaintx.rs [hunk 5]
```diff
@@ -773,7 +795,15 @@ impl<ChannelSigner: WriteableEcdsaChannelSigner> OnchainTxHandler<ChannelSigner>
...
 							//TODO: recompute soonest_timelock to avoid wasting a bit on fees
 							if at_least_one_drop {
-								bump_candidates.insert(first_claim_txid_height.0.clone(), request.clone());
+								bump_candidates.insert(*package_id, request.clone());
+								// If we have any pending claim events for the request being updated
+								// that have yet to be consumed, we'll remove them since they will
+								// end up producing an invalid transaction by double spending
+								// input(s) that already have a confirmed spend. If such spend is
+								// reorged out of the chain, then we'll attempt to re-spend the
+								// inputs once we see it.
```
Comment (ariard) [inline lightning/src/chain/onchaintx.rs:804]: We still re-generate adequate `ClaimEvent` L965 in `block_disconnected` though I'm not sure if we have test coverage for both post-anchor bump HTLC/bump commitment ? Or at least doesn't seem so from a look in `monitor_tests`, `reorg_tests` and `functional_tests`. Can be added in future PRs to avoid delaying more this one.

**Review by ariard — COMMENTED** [review]
Thanks for adding the test coverage extension on correct punishment of multiple revoked channels!

File: lightning/src/chain/onchaintx.rs [hunk 6]
```diff
@@ -846,10 +885,15 @@ impl<ChannelSigner: WriteableEcdsaChannelSigner> OnchainTxHandler<ChannelSigner>
 					#[cfg(anchors)]
 					OnchainClaim::Event(claim_event) => {
 						log_info!(logger, "Yielding RBF-bumped onchain event to spend inputs {:?}", request.outpoints());
-						self.pending_claim_events.insert(*first_claim_txid, claim_event);
+						if let Some((existing_claim_idx, _)) = self.pending_claim_events.iter().enumerate()
```
Comment (arik-so) [inline lightning/src/chain/onchaintx.rs:888]: Should this stop at the first find? Also, what's the utility of this rather verbose expression rather than a retain for inequality?

File: lightning/src/chain/onchaintx.rs [hunk 7]
```diff
@@ -922,7 +966,12 @@ impl<ChannelSigner: WriteableEcdsaChannelSigner> OnchainTxHandler<ChannelSigner>
 					#[cfg(anchors)]
 					OnchainClaim::Event(claim_event) => {
 						log_info!(logger, "Yielding onchain event after reorg to spend inputs {:?}", request.outpoints());
-						self.pending_claim_events.insert(_first_claim_txid_height.0, claim_event);
+						if let Some((existing_claim_idx, _)) = self.pending_claim_events.iter().enumerate()
```
Comment (arik-so) [inline lightning/src/chain/onchaintx.rs:969]: same here. What's the benefit of a) stopping on the first find and b) not using retain?

File: lightning/src/chain/onchaintx.rs [hunk 8]
```diff
@@ -773,7 +804,21 @@ impl<ChannelSigner: WriteableEcdsaChannelSigner> OnchainTxHandler<ChannelSigner>
...
-								bump_candidates.insert(first_claim_txid_height.0.clone(), request.clone());
+								bump_candidates.insert(*package_id, request.clone());
+								// If we have any pending claim events for the request being updated
+								// that have yet to be consumed, we'll remove them since they will
+								// end up producing an invalid transaction by double spending
+								// input(s) that already have a confirmed spend. If such spend is
+								// reorged out of the chain, then we'll attempt to re-spend the
+								// inputs once we see it.
+								#[cfg(anchors)] {
+									#[cfg(debug_assertions)] {
```
Comment (arik-so) [inline lightning/src/chain/onchaintx.rs:815]: I apologize for being dumb, but what's the benefit of cfg(debug_assertions) with an assert over debug_assert with no config flag? Is it just an optimization to not do the counting?

File: lightning/src/ln/functional_test_utils.rs [hunk 9]
```diff
@@ -1239,24 +1239,23 @@ macro_rules! check_warn_msg {
...
-		},
-		_ => panic!("Unexpected event"),
-	}
-	if with_error_msg {
-		match msg_events[1] {
+	assert_eq!(msg_events.len(), if with_error_msg { num_channels * 2 } else { num_channels });
+	msg_events.into_iter().filter_map(|msg_event| {
+		match msg_event {
+			MessageSendEvent::BroadcastChannelUpdate { ref msg } => {
+				assert_eq!(msg.contents.flags & 2, 2);
```
Comment (arik-so) [inline lightning/src/ln/functional_test_utils.rs:1248]: if there is an overlap between people working on this code and people who haven't memorized the meaning of the second bit in the channel update message, this needs a comment. However, I'm not sure such an overlap exists.

Comment (arik-so) [reply lightning/src/ln/functional_test_utils.rs:1248]: or… we could use a constant instead of a magic 2. I know it's just used in tests, but given it's a test utility and not an individual test, might be some merit to not forcing readers to look up the spec. I won't block the PR over this though.

**Review by arik-so — APPROVED** [review]
